src/ch/ch_hotplug.c

```diff
@@ -0,0 +1,194 @@
+#include "ch_hotplug.h"
+#include "ch_monitor.h"
+#include "ch_process.h"
+#include "domain_audit.h"
+
+#define VIR_FROM_THIS VIR_FROM_CH
+
+/**
+ * virCHDomainSelectHotplugVcpuEntities:
+ *
+ * @def: domain definition
+ * @nvcpus: target vcpu count
+ *
+ * Validate vcpu entities will be able to be enabled/disabled.
+ *
+ * Returns the true if the nvcpus can be set, false if not.
+ */
+static bool
+virCHDomainSelectHotplugVcpuEntities(virDomainDefPtr def,
+                                     unsigned int nvcpus)
+{
+    bool ret = false;
+    virDomainVcpuDefPtr vcpu;
+    unsigned int maxvcpus = virDomainDefGetVcpusMax(def);
+    unsigned int curvcpus = virDomainDefGetVcpus(def);
+    ssize_t i;
+
+    if (nvcpus > curvcpus) {
+        for (i = 0; i < maxvcpus && curvcpus < nvcpus; i++) {
+            vcpu = virDomainDefGetVcpu(def, i);
+
+            if (vcpu->online)
+                continue;
+
+            curvcpus++;
+
+            if (curvcpus > nvcpus) {
+                virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                               _("target vm vcpu granularity does not allow the "
+                                 "desired vcpu count"));
+                goto error;
+            }
+        }
+    } else {
+        for (i = maxvcpus - 1; i >= 0 && curvcpus > nvcpus; i--) {
+            vcpu = virDomainDefGetVcpu(def, i);
+
+            if (!vcpu->online)
+                continue;
+
+            curvcpus--;
+
+            if (curvcpus < nvcpus) {
+                virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                               _("target vm vcpu granularity does not allow the "
+                                 "desired vcpu count"));
+                goto error;
+            }
+        }
+    }
+
+    if (curvcpus != nvcpus) {
+        virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                       _("failed to find appropriate hotpluggable vcpus to "
+                         "reach the desired target vcpu count"));
+        goto error;
+    }
+
+    ret = true;
+
+ error:
+    return ret;
+}
+
+static int
+virCHDomainHotplugSetVcpus(virCHDriverPtr driver,
+                           virCHDriverConfigPtr cfg,
+                           virDomainObjPtr vm,
+                           unsigned int nvcpus)
+{
+    int oldvcpus = virDomainDefGetVcpus(vm->def);
+    virCHDomainObjPrivatePtr priv = vm->privateData;
+    virCHMonitorPtr mon = priv->monitor;
+    int ret = -1;
+
+    ret = virCHMonitorResizeCPU(mon, nvcpus);
+
+    virDomainAuditVcpu(vm, oldvcpus, nvcpus, "update", ret == 0);
+
+    if (ret < 0)
+        goto error;
+
+    for (int i = 0; i < nvcpus; i++) {
+        virDomainVcpuDefPtr vcpuinfo = virDomainDefGetVcpu(vm->def, i);
+        vcpuinfo->online = true;
+    }
+
+    if (virCHMonitorRefreshThreadInfo(mon) < 0)
+        goto error;
+
+    if (virCHProcessSetupVcpus(vm) < 0)
+        goto error;
+
+    if (virCHDomainValidateVcpuInfo(vm) < 0)
+        goto error;
```

Comment:
I feel that we need not have this logic here. When there is a change in the vcpu threads, we might get a monitor event and the monitor thread might take care of it. Could you please give it a try?

Reply:
I'm sorry I am not quite sure what the change you are looking for here is. The first part of your comment I took to removing the `virCHDomainValidateVcpuInfo` function entirely initially. But then reading your comment about moving the `virCHDomainValidateVcpuInfo` function I think maybe you mean removing the `virCHMonitorRefreshThreadInfo` and `virCHProcessSetupVcpus` calls. 

I'm hesitant remove those two calls because if they fail, the failure would be caught at the point of hotplug execution here. Otherwise the failure would be caught sometime later in a more asynchronous check that could be harder to associate with the hotplug api call.

Thoughts?

Reply:
Sorry for not being clear. I was mentioning that `virCHMonitorRefreshThreadInfo` and `virCHProcessSetupVcpus` are being done in `virCHProcessSetupThreads` whenever we get a lifecycle event about vm or vcpu. So it makes sense to let the monitor thread take care of this update and to avoid duplicating it. You have a valid point that doing it synchronously will help to catch Hotplug errors. But there is always a chance for a race here: If we call `virCHMonitorRefreshThreadInfo` before the threads are populated in procfs, then we get a false alarm here. We would not be having a race in the monitor event handler.

I am fine with duplicating the logic here. Maybe you can add `virCHDomainValidateVcpuInfo` to the `virCHProcessSetupThreads` as well so that we validate the vcpu thread information every time there is a change.

Reply:
Thinking a little more about this, the `virCHDomainHotplugSetVcpus` sets the expected vcpus online (which it knows via `nvcpus`). Then `virCHDomainValidateVcpuInfo` is called to ensure the tids that were read from proc are attached to those newly marked online cpus (and only the marked online cpus). If I add `virCHDomainValidateVcpuInfo` to `virCHProcessSetupThreads`, it would use the last set cpu online values in the `vm->def` which as long as it was from the corresponding hotplug would be fine but isn't that racy too?

It seems we could potentially always be subject to some levels of race condition around when proc updates even if we only read after cloud-hypervisor sends us a message to check.

In the hopes of avoiding stacking potential races, I'd not duplicate the calls and just either move the update and validation to the `virCHProcessSetupThreads` call or keep it as is. Should the `virCHProcessSetupThreads` also be making a `virDomainObjSave` call (and maybe remove the call from `virCHDomainHotplugSetVcpus`)?

Reply:
I agree about not duplicating the code. And it makes sense to call `virDomainObjSave` in `virCHProcessSetupThreads`. So, I feel we have two options:
1. Remove lines 120 to 130, add `virDomainObjSave` to `virCHProcessSetupThreads` and call `virCHProcessSetupThreads` here in this `virCHDomainHotplugSetVcpus`
2. Remove lines 120 to 130, add `virDomainObjSave` to `virCHProcessSetupThreads` and rely on the event monitor thread to update the vcpu pid information.

Both are fine with me.

Reply:
Went with option 2, though as noted in my last commit there is an issue with detecting removing vcpus.